src/realm/util/file_mapper.cpp

```diff
@@ -87,6 +93,171 @@ size_t round_up_to_page_size(size_t size) noexcept
     return (size + page_size() - 1) & ~(page_size() - 1);
 }
 
+// Support for logging memory mappings
+std::mutex mmap_log_mutex;
+struct MMapEntry {
+    enum class Type { Reserved, File, Memory };
+    Type type;
+    size_t offset;
+    char* start;
+    char* end;
+    std::string path;
+    const char* cause;
+};
+
+std::map<void*, MMapEntry> all_mappings;
+static std::string dump_logged_mappings();
+
+static void remove_logged_mapping(char* start, size_t size)
+{
+    REALM_ASSERT(start != (char*)-1);
+    REALM_ASSERT(size != 0);
+    char* end = start + size;
+    auto it = all_mappings.lower_bound(start);
+    // previous entry may overlap, so must be included in scan:
+    if (it != all_mappings.begin()) {
+        --it;
+    }
+    // scan for and adjust/remove entries which overlaps range:
+    while (it != all_mappings.end() && it->second.start < end) {
+        // handle case where the range "punches a hole" in existing entry
+        if (it->second.start < start && end < it->second.end) {
+            // collect info for part after hole
+            size_t offset = 0;
+            if (it->second.type == MMapEntry::Type::File) {
+                offset = it->second.offset + end - it->second.start;
+            }
+            REALM_ASSERT((offset % 4096) == 0);
+            REALM_ASSERT(((size_t)end % 4096) == 0);
+            // reduce size of entry to part before hole
+            auto old_end = it->second.end;
+            it->second.end = start;
+            // create new entry for part after hole
+            MMapEntry entry{it->second.type, offset, end, old_end, it->second.path};
+            all_mappings[end] = entry;
+            // aaaaand we're done
```

Comment:
all_mappings.emplace(end, MMapEntry{it->second.type, offset, end, old_end, it->second.path});